File: xkcd/src/main/java/xyz/jienan/xkcd/model/persist/BoxManager.java
package xyz.jienan.xkcd.model.persist;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.android.gms.common.util.NumberUtils;

import java.util.Collections;
import java.util.List;

import io.objectbox.Box;
import io.objectbox.query.Query;
import io.objectbox.query.QueryBuilder;
import xyz.jienan.xkcd.XkcdApplication;
import xyz.jienan.xkcd.model.WhatIfArticle;
import xyz.jienan.xkcd.model.WhatIfArticle_;
import xyz.jienan.xkcd.model.XkcdPic;
import xyz.jienan.xkcd.model.XkcdPic_;

import static xyz.jienan.xkcd.Const.PREF_WHAT_IF_SEARCH_ALL;
import static xyz.jienan.xkcd.Const.PREF_WHAT_IF_SEARCH_INCLUDE_READ;

public class BoxManager {

    private static BoxManager boxManager;
    private Box<XkcdPic> xkcdBox;
    private Box<WhatIfArticle> whatIfBox;

    private BoxManager() {
        xkcdBox = XkcdApplication.getInstance().getBoxStore().boxFor(XkcdPic.class);
        whatIfBox = XkcdApplication.getInstance().getBoxStore().boxFor(WhatIfArticle.class);
    }

    public static BoxManager getInstance() {
        if (boxManager == null) {
            boxManager = new BoxManager();
        }
        return boxManager;
    }

    /********** xkcd **********/

    @Nullable
    public XkcdPic getXkcd(long index) {
        return xkcdBox.get(index);
    }

    public void saveXkcd(XkcdPic xkcdPic) {
        xkcdBox.put(xkcdPic);
    }

    @Nullable
    public XkcdPic likeXkcd(long index) {
        final XkcdPic xkcdPic = xkcdBox.get(index);
        if (xkcdPic != null) {
            xkcdPic.hasThumbed = true;
            xkcdBox.put(xkcdPic);
        }
        return xkcdPic;
    }

    public XkcdPic favXkcd(long index, boolean isFav) {
        final XkcdPic xkcdPic = xkcdBox.get(index);
        if (xkcdPic != null) {
            xkcdPic.isFavorite = isFav;
            xkcdBox.put(xkcdPic);
        }
        return xkcdPic;
    }

    @NonNull
    public List<XkcdPic> getXkcdInRange(long start, long end) {
        final Query<XkcdPic> query = xkcdBox.query().between(XkcdPic_.num, start, end).build();
        return query.find();
    }

    @NonNull
    public List<XkcdPic> getValidXkcdInRange(long start, long end) {
        final Query<XkcdPic> query = xkcdBox.query()
                .between(XkcdPic_.num, start, end)
                .and().greater(XkcdPic_.width, 0).build();
        return query.find();
    }

    @NonNull
    public List<XkcdPic> getFavXkcd() {
        final Query<XkcdPic> queryFav = xkcdBox.query().equal(XkcdPic_.isFavorite, true).build();
        return queryFav.find();
    }

    @NonNull
    public List<XkcdPic> updateAndSave(@NonNull List<XkcdPic> xkcdPics) {
        for (XkcdPic pic : xkcdPics) {
            XkcdPic xkcdPicInBox = xkcdBox.get(pic.num);
            if (xkcdPicInBox != null) {
                pic.isFavorite = xkcdPicInBox.isFavorite;
                pic.hasThumbed = xkcdPicInBox.hasThumbed;
                if (pic.width == 0 && xkcdPicInBox.width != 0) {
                    pic.width = xkcdPicInBox.width;
                    pic.height = xkcdPicInBox.height;
                }
            }
        }
        xkcdBox.put(xkcdPics);
        return xkcdPics;
    }

    @NonNull
    public XkcdPic updateAndSave(@NonNull XkcdPic xkcdPic) {
        return updateAndSave(Collections.singletonList(xkcdPic)).get(0);
    }

    /********** what if **********/

    @Nullable
    public List<WhatIfArticle> getWhatIfArchive() {
        return whatIfBox.getAll();
    }

    @Nullable
    public WhatIfArticle getWhatIf(long index) {
        return whatIfBox.get(index);
    }

    @NonNull
    public List<WhatIfArticle> updateAndSaveWhatIf(@NonNull List<WhatIfArticle> whatIfArticles) {
        for (WhatIfArticle article : whatIfArticles) {
            WhatIfArticle articleInBox = whatIfBox.get(article.num);
            if (articleInBox != null && articleInBox.content != null) {
                whatIfArticles.set((int) article.num - 1, articleInBox);
            }
        }
        whatIfBox.put(whatIfArticles);
        return whatIfArticles;
    }

    public WhatIfArticle updateAndSaveWhatIf(long index, String content) {
        final WhatIfArticle article = whatIfBox.get(index);
        if (article != null) {
            article.content = content;
            whatIfBox.put(article);
        }
        return article;
    }

    @NonNull
    public List<WhatIfArticle> searchWhatIf(@NonNull String query, String option) {
        QueryBuilder<WhatIfArticle> builder = whatIfBox.query().contains(WhatIfArticle_.title, query);
        if (NumberUtils.isNumeric(query)) {
            builder = builder.or().equal(WhatIfArticle_.num, Long.valueOf(query));
        }
        switch (option) {
            case PREF_WHAT_IF_SEARCH_INCLUDE_READ:
                builder = builder.or()
                        .contains(WhatIfArticle_.content, query)
                        .and()
                        .equal(WhatIfArticle_.hasThumbed, true)
                        .or()
                        .equal(WhatIfArticle_.isFavorite, true);
                break;
            case PREF_WHAT_IF_SEARCH_ALL:
                builder = builder.or()
                        .contains(WhatIfArticle_.content, query);
                break;
        }
        return builder.build().find();
    }

    @Nullable
    public WhatIfArticle likeWhatIf(long index) {
        final WhatIfArticle article = whatIfBox.get(index);
        if (article != null) {
            article.hasThumbed = true;
            whatIfBox.put(article);
        }
        return article;
    }

    public WhatIfArticle favWhatIf(long index, boolean isFav) {
        final WhatIfArticle article = whatIfBox.get(index);
        if (article != null) {
            article.isFavorite = isFav;
            whatIfBox.put(article);
        }
        return article;
    }

    public List<WhatIfArticle> getFavWhatIf() {
        final Query<WhatIfArticle> queryFav = whatIfBox.query().equal(WhatIfArticle_.isFavorite, true).build();
        return queryFav.find();
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/model/persist/BoxManager.kt
package xyz.jienan.xkcd.model.persist

import io.objectbox.Box
import io.objectbox.BoxStore
import org.jsoup.internal.StringUtil
import xyz.jienan.xkcd.Const.PREF_WHAT_IF_SEARCH_ALL
import xyz.jienan.xkcd.Const.PREF_WHAT_IF_SEARCH_INCLUDE_READ
import xyz.jienan.xkcd.model.*
import java.util.*


object BoxManager {

    lateinit var xkcdBox: Box<XkcdPic>
        private set

    lateinit var whatIfBox: Box<WhatIfArticle>
        private set

    private lateinit var extraBox: Box<ExtraComics>

    fun init(boxStore: BoxStore?) {
        xkcdBox = boxStore!!.boxFor(XkcdPic::class.java)
        whatIfBox = boxStore.boxFor(WhatIfArticle::class.java)
        extraBox = boxStore.boxFor(ExtraComics::class.java)
    }

    val favXkcd: List<XkcdPic>
        get() {
            val queryFav = xkcdBox.query().equal(XkcdPic_.isFavorite, true).build()
            return queryFav.find()
        }

    val untouchedComicsList: List<XkcdPic>
        get() {
            val query = xkcdBox.query().notEqual(XkcdPic_.isFavorite, true)
                    .and().notEqual(XkcdPic_.hasThumbed, true).build()
            return query.find()
        }

    /********** what if  */

    val whatIfArchive: List<WhatIfArticle>?
        get() = whatIfBox.all

    val favWhatIf: List<WhatIfArticle>
        get() {
            val queryFav = whatIfBox.query().equal(WhatIfArticle_.isFavorite, true).build()
            return queryFav.find()
        }

    val untouchedArticleList: List<WhatIfArticle>
        get() {
            val query = whatIfBox.query().notEqual(WhatIfArticle_.isFavorite, true)
                    .and().notEqual(WhatIfArticle_.hasThumbed, true).build()
            return query.find()
        }

    val extraList: List<ExtraComics>
        get() = if (extraBox.isEmpty) ArrayList() else extraBox.all

    /********** xkcd  */

    fun getXkcd(index: Long): XkcdPic? {
        return xkcdBox.get(index)
    }

    fun saveXkcd(xkcdPic: XkcdPic) {
        xkcdBox.put(xkcdPic)
    }

    fun likeXkcd(index: Long): XkcdPic? {
        val xkcdPic = xkcdBox.get(index)
        if (xkcdPic != null) {
            xkcdPic.hasThumbed = true
            xkcdBox.put(xkcdPic)
        }
        return xkcdPic
    }

    fun favXkcd(index: Long, isFav: Boolean): XkcdPic? {
        val xkcdPic = xkcdBox.get(index)
        if (xkcdPic != null) {
            xkcdPic.isFavorite = isFav
            xkcdBox.put(xkcdPic)
        }
        return xkcdPic
    }

    fun getXkcdInRange(start: Long, end: Long): List<XkcdPic> {
        val query = xkcdBox.query().between(XkcdPic_.num, start, end).build()
        return query.find()
    }

    fun getValidXkcdInRange(start: Long, end: Long): List<XkcdPic> {
        val query = xkcdBox.query()
                .between(XkcdPic_.num, start, end)
                .and().greater(XkcdPic_.width, 0).build()
        return query.find()
    }

    fun updateAndSave(xkcdPics: List<XkcdPic>): List<XkcdPic> {
        for (pic in xkcdPics) {
            val xkcdPicInBox = xkcdBox.get(pic.num)
            if (xkcdPicInBox != null) {
                pic.isFavorite = xkcdPicInBox.isFavorite
                pic.hasThumbed = xkcdPicInBox.hasThumbed
                if (pic.width == 0 && xkcdPicInBox.width != 0) {
                    pic.width = xkcdPicInBox.width
                    pic.height = xkcdPicInBox.height
                }
            }
        }
        xkcdBox.put(xkcdPics)
        return xkcdPics
    }

    fun updateAndSave(xkcdPic: XkcdPic): XkcdPic = updateAndSave(listOf(xkcdPic))[0]

    fun getWhatIf(index: Long): WhatIfArticle? = whatIfBox.get(index)

    fun updateAndSaveWhatIf(whatIfArticles: MutableList<WhatIfArticle>): List<WhatIfArticle> {
        for (article in whatIfArticles) {
            val articleInBox = whatIfBox.get(article.num)
            if (articleInBox?.content != null) {
                whatIfArticles[article.num.toInt() - 1] = articleInBox
            }
        }
        whatIfBox.put(whatIfArticles)
        return whatIfArticles
    }

    fun updateAndSaveWhatIf(index: Long, content: String): WhatIfArticle? {
        val article = whatIfBox.get(index)
        if (article != null) {
            article.content = content
            whatIfBox.put(article)
        }
        return article
    }

    fun searchWhatIf(query: String, option: String): MutableList<WhatIfArticle> {
        var builder = whatIfBox.query().contains(WhatIfArticle_.title, query)
        if (StringUtil.isNumeric(query)) {
            builder = builder.or().equal(WhatIfArticle_.num, query.toLong())
        }
        when (option) {
            PREF_WHAT_IF_SEARCH_INCLUDE_READ -> {
                // ObjectBox is not good at such query filters
                val listInTitle = builder.build().find()
                val listInReadContent = whatIfBox.query()
                        .contains(WhatIfArticle_.content, query)
                        .build()
                        .find()
                        .filter { it.hasThumbed || it.isFavorite }
                return listInTitle.union(listInReadContent).toMutableList()
            }
            PREF_WHAT_IF_SEARCH_ALL -> builder = builder.or()
                    .contains(WhatIfArticle_.content, query)
        }
        return builder.build().find()
    }

    fun likeWhatIf(index: Long): WhatIfArticle? {
        val article = whatIfBox.get(index)
        if (article != null) {
            article.hasThumbed = true
            whatIfBox.put(article)
        }
        return article
    }

    fun favWhatIf(index: Long, isFav: Boolean): WhatIfArticle? {
        val article = whatIfBox.get(index)
        if (article != null) {
            article.isFavorite = isFav
            whatIfBox.put(article)
        }
        return article
    }

    /********** extra  */

    fun saveExtras(extraComics: List<ExtraComics>) {
        extraBox.put(extraComics)
    }

    fun getExtra(index: Int): ExtraComics = extraBox.get(index.toLong())

    fun loadExtraExplain(url: String) =
            extraBox.query().equal(ExtraComics_.explainUrl, url).build().findFirst()?.explainContent

    fun updateExtra(url: String, explainContent: String) {
        val extraComics = extraBox.query().equal(ExtraComics_.explainUrl, url).build().findFirst()
        if (extraComics != null) {
            extraComics.explainContent = explainContent
            extraBox.put(extraComics)
        }
    }
}
